flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSink.java

```diff
@@ -0,0 +1,696 @@
+    this.targetDataFileSize = targetDataFileSize;
+    this.dataFileFormat = dataFileFormat;
+  }
+
+  @Override
+  public DataStream<RowData> addPreWriteTopology(DataStream<RowData> inputDataStream) {
+    return distributeDataStream(
+        inputDataStream,
+        equalityFieldIds,
+        table.spec(),
+        table.schema(),
+        flinkRowType,
+        distributionMode,
+        equalityFieldColumns);
+  }
+
+  @Override
+  public StreamWriter createWriter(InitContext context) {
+    return restoreWriter(context, null);
+  }
+
+  @Override
+  public StreamWriter restoreWriter(
+      InitContext context, Collection<StreamWriterState> recoveredState) {
+    RowDataTaskWriterFactory taskWriterFactory =
+        new RowDataTaskWriterFactory(
+            SerializableTable.copyOf(table),
+            flinkRowType,
+            targetDataFileSize,
+            dataFileFormat,
+            equalityFieldIds,
+            upsertMode);
+    StreamWriter streamWriter =
+        new StreamWriter(
+            table.name(),
+            taskWriterFactory,
+            context.getSubtaskId(),
+            context.getNumberOfParallelSubtasks());
+    if (recoveredState == null) {
+      return streamWriter;
+    }
+
+    return streamWriter.restoreWriter(recoveredState);
+  }
+
+  @Override
+  public SimpleVersionedSerializer<StreamWriterState> getWriterStateSerializer() {
+    return new StreamWriterStateSerializer();
+  }
+
+  @Override
+  public DataStream<CommittableMessage<FilesCommittable>> addPreCommitTopology(
+      DataStream<CommittableMessage<FilesCommittable>> writeResults) {
+    return writeResults
+        .map(
+            new RichMapFunction<
+                CommittableMessage<FilesCommittable>, CommittableMessage<FilesCommittable>>() {
+              @Override
+              public CommittableMessage<FilesCommittable> map(
+                  CommittableMessage<FilesCommittable> message) {
+                if (message instanceof CommittableWithLineage) {
+                  CommittableWithLineage<FilesCommittable> committableWithLineage =
+                      (CommittableWithLineage<FilesCommittable>) message;
+                  FilesCommittable committable = committableWithLineage.getCommittable();
+                  committable.checkpointId(committableWithLineage.getCheckpointId().orElse(-1));
+                  committable.subtaskId(committableWithLineage.getSubtaskId());
+                  committable.jobID(getRuntimeContext().getJobId().toString());
+                }
+                return message;
+              }
+            })
+        .uid(uidPrefix + "-pre-commit-topology")
+        .global();
+  }
+
+  @Override
+  public Committer<FilesCommittable> createCommitter() {
+    return committer != null
+        ? committer
+        : new FilesCommitter(tableLoader, overwrite, snapshotProperties, workerPoolSize);
+  }
+
+  public void setCommitter(Committer<FilesCommittable> committer) {
```

Comment:
why is this method needed?

Reply:
We don’t usually use `set` for our own function names. Is this to make this serializable? If not, I’d skip the `set` verb.

Reply:
Just for ut. Maybe add `@VisibleForTesting` for it?

Reply:
yeah. adding the annotation would help. also if the unit test is in the same package, we can also make it package private.